### ISSUE TITLE
Make BrokerSetAwareGoal capable of doing partition coloring

### ISSUE THREAD

**Issue by CCisGG** [linkedin/cruise-control#1862]
Recently AWS has make contributions to CC and create this BrokerSetAwareGoal in: https://github.com/linkedin/cruise-control/pull/1809. This goal will ensure each topic (all its partitions and replicas) only lives within a single broker set.

Today we have a need to do partition level coloring, meaning a partition and all its replicas live within a broker set. Different partitions of a topic doesn't need to be in the same color.  With the current BrokerSetAwareGoal, it is not allowed. I'm thinking about to add partition coloring by either update the BrokerSetAwareGoal, or make BrokerSetAwareGoal easily extendable to achieve it.

Basically there are 2 things we need to change:
1. For "ensureBrokerSetAware" method, we need to make it accept partition level coloring.
2. We need to remove the _mustHaveTopicLeadersPerBroker affect. In topic coloring, the rebalance should ignore those topics; while in partition coloring, the rebalance should not ignore those topics.
